Make some minor improvements for LastValueCache

llucax opened the pull request:
- **Add a `unique_id` to `LatestValueCache`**
- **Add `__str__` and `__repr__` to `LatestValueCache`**
- **Cancel `LatestValueCache`'s task if deleted**
- **Add `@override` to `PowerDistributingActor`**

llucax committed:
Add a `unique_id` to `LatestValueCache`

When debugging and having some task in the stack coming from a
`LastValueCache` it is very difficult to identify which one is it.

This commit add a new `unique_id` that's used to create the task name.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add `__str__` and `__repr__` to `LatestValueCache`

The `str` just returns the value, and `repr` gets all the info,
including the new `unique_id`.

We also implement `__str__` for `_Sentinel` to return a more meaningful
string, since it will be shown in the `LatestValueCache` string
representation when no value has been received yet.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add `@override` to `PowerDistributingActor`

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented:
Enabled auto-merge.

shsms commented on src/frequenz/sdk/_internal/_channels.py:
```diff
@@ -91,3 +91,14 @@ async def _run(self) -> None:
+
+    def __str__(self) -> str:
+        """Return a string representation of the cache."""
+        return str(self._latest_value)
```
Maybe the docstring should say the latest value in the cache instead?

llucax replied:
OK, I tend to use just a generic docstring for `__str__` and `__repr__` but maybe in this case it is worth to be more explicit.

shsms commented on src/frequenz/sdk/_internal/_channels.py:
```diff
@@ -92,6 +92,10 @@ async def stop(self) -> None:
 
+    def __del__(self) -> None:
+        """Stop the cache when it is deleted."""
+        self._task.cancel()
```
Best to check if the event loop is still running before calling cancel, else, we might get more warning spam in tests.

llucax replied:
Oh, really? Actually I added this because I was getting warnings in tests about the last value cache tasks being pending when the loop was destroyed. But I debugged the issue and the tasks were properly finished, it seems it might be a bug in pytest-async, I believe we had some similar issues in the past.

Maybe I should remove this and that's it, I think according to Python specs this should not be necessary, and tasks should be cancelled (or discarded) when the even loop is closed, and dangling tasks will also be removed from the loop and finished, it might only happen that this is delayed until the GC runs if the reference was part of a reference loop.

I'd say until we are completely sure this is really necessary we just don't add it.

llucax replied:
Just to add more context, this is the issue that made me try this:

* https://github.com/frequenz-floss/frequenz-sdk-python/pull/962#issuecomment-2182264984

llucax commented:
Updated to improve the `__str__` docstring and removed the commit adding `__del__`. I also implemented `__str__` for `_Sentinel` so we get a more meaningful string when the cache didn't receive any value yet.
